pkg/cache/cache_test.go

```diff
@@ -93,6 +121,76 @@ var _ = Describe("Multi-Namespace Informer Cache", func() {
 var _ = Describe("Informer Cache without DeepCopy", func() {
 	CacheTest(cache.New, cache.Options{UnsafeDisableDeepCopyByObject: cache.DisableDeepCopyByObject{cache.ObjectAll{}: true}})
 })
+var _ = Describe("Cache with selectors", func() {
+	defer GinkgoRecover()
+	var (
+		informerCache       cache.Cache
+		informerCacheCtx    context.Context
+		informerCacheCancel context.CancelFunc
+	)
+
+	BeforeEach(func() {
+		informerCacheCtx, informerCacheCancel = context.WithCancel(context.Background())
+		Expect(cfg).NotTo(BeNil())
+		cl, err := client.New(cfg, client.Options{})
+		Expect(err).NotTo(HaveOccurred())
+		err = ensureNamespace(testNamespaceOne, cl)
+		Expect(err).NotTo(HaveOccurred())
+		err = ensureNamespace(testNamespaceTwo, cl)
+		Expect(err).NotTo(HaveOccurred())
+		for idx, namespace := range []string{testNamespaceOne, testNamespaceTwo} {
+			_ = createSA("test-sa-"+strconv.Itoa(idx), namespace, cl)
+			_ = createSvc("test-svc-"+strconv.Itoa(idx), namespace, cl)
+		}
+
+		opts := cache.Options{
+			SelectorsByObject: cache.SelectorsByObject{
+				&corev1.ServiceAccount{}:             {Field: fields.OneTermEqualSelector("metadata.namespace", testNamespaceOne)},
+				&cache.SelectorsByObjectDefaultKey{}: cache.ObjectSelector{Field: fields.OneTermEqualSelector("metadata.namespace", testNamespaceTwo)},
+			},
```

Comment:
Looking at how this is being used, if I understood what we're trying to do here correctly, we can summarize it as:
- For objects of type ServiceAccount, always cache only the objects in namespace 1
- For any other object, cache objects in namespace 2

If that's the case, I'm not super sure if I like this APIs, it's confusing that the fallback is also a "selector by object", which seems like a wildcard, although it could be confusing to users.

Have we thought instead of adding another option, like `DefaultSelector`, or similar?

Reply:
@vincepri yes, see the discussion here: https://github.com/kubernetes-sigs/controller-runtime/pull/1710#issuecomment-961179320

Reply:
I would suggest to not add and deprecate any magic key in favor of a different field, it makes a nicer UX and easier understanding of the codebase. Without this test/example, it wouldn't have been clear what the PR was trying to achieve and how.

Reply:
Okay, reverted it back to add a top-level `DefaultSelector` field, ptal.